contracts/tokenbridge/ethereum/L1AtomicTokenBridgeCreator.sol

```diff
@@ -0,0 +1,454 @@
+            router,
+            inbox,
+            keccak256(type(ClonableBeaconProxy).creationCode),
+            canonicalL2BeaconProxyFactoryAddress
+        );
+
+        return address(standardGateway);
+    }
+
+    function _deployL1CustomGateway(address proxyAdmin, address router, address inbox, address owner)
+        internal
+        returns (address)
+    {
+        L1CustomGateway customGateway = L1CustomGateway(
+            address(
+                new TransparentUpgradeableProxy{
+                    salt: _getL1Salt(OrbitSalts.L1_CUSTOM_GATEWAY, inbox)
+                }(address(customGatewayTemplate), proxyAdmin, bytes(""))
+            )
+        );
+
+        customGateway.initialize(getCanonicalL2CustomGatewayAddress(), router, inbox, owner);
+
+        return address(customGateway);
+    }
+
+    function _deployL1WethGateway(address proxyAdmin, address router, address inbox) internal returns (address) {
+        L1WethGateway wethGateway = L1WethGateway(
+            payable(
+                address(
+                    new TransparentUpgradeableProxy{
+                        salt: _getL1Salt(OrbitSalts.L1_WETH_GATEWAY, inbox)
+                    }(address(wethGatewayTemplate), proxyAdmin, bytes(""))
+                )
+            )
+        );
+
+        wethGateway.initialize(getCanonicalL2WethGatewayAddress(), router, inbox, l1Weth, getCanonicalL2WethAddress());
+
+        return address(wethGateway);
+    }
+
+    function _deployL2Factory(address inbox, uint256 maxSubmissionCost, uint256 maxGas, uint256 gasPriceBid) internal {
+        // encode L2 factory bytecode
+        bytes memory deploymentData = _creationCodeFor(l2TokenBridgeFactoryTemplate.code);
+
+        uint256 value = maxSubmissionCost + maxGas * gasPriceBid;
+        IInbox(inbox).createRetryableTicket{value: value}(
+            address(0), 0, maxSubmissionCost, msg.sender, msg.sender, maxGas, gasPriceBid, deploymentData
+        );
+    }
+
+    function _deployL2Contracts(
+        address l1Router,
+        address l1StandardGateway,
+        address l1CustomGateway,
+        address l1WethGateway,
+        address inbox,
+        uint256 maxSubmissionCost,
+        uint256 maxGas,
+        uint256 gasPriceBid
+    ) internal {
+        address l2ProxyAdminOwner = _getRollupOwner(inbox);
+        bytes memory data = abi.encodeWithSelector(
+            L2AtomicTokenBridgeFactory.deployL2Contracts.selector,
+            _creationCodeFor(l2RouterTemplate.code),
+            _creationCodeFor(l2StandardGatewayTemplate.code),
+            _creationCodeFor(l2CustomGatewayTemplate.code),
+            _creationCodeFor(l2WethGatewayTemplate.code),
+            _creationCodeFor(l2WethTemplate.code),
+            l1Router,
+            l1StandardGateway,
+            l1CustomGateway,
+            l1WethGateway,
+            l1Weth,
+            getCanonicalL2StandardGatewayAddress(),
+            l2ProxyAdminOwner
+        );
```

Comment:
recently learnt that you can use abi.encodeCall which is safer since it check types
```suggestion
        bytes memory data = abi.encodeCall(
            L2AtomicTokenBridgeFactory.deployL2Contracts,
            (_creationCodeFor(l2RouterTemplate.code),
            _creationCodeFor(l2StandardGatewayTemplate.code),
            _creationCodeFor(l2CustomGatewayTemplate.code),
            _creationCodeFor(l2WethGatewayTemplate.code),
            _creationCodeFor(l2WethTemplate.code),
            l1Router,
            l1StandardGateway,
            l1CustomGateway,
            l1WethGateway,
            l1Weth,
            getCanonicalL2StandardGatewayAddress(),
            l2ProxyAdminOwner)
        );
```

Reply:
nice, done